src/extractor/edge_based_graph_factory.cpp

```diff
@@ -436,34 +434,51 @@ void EdgeBasedGraphFactory::GenerateEdgeExpandedEdges(
 
                 const int32_t turn_penalty =
                     scripting_environment.GetTurnPenalty(180. - turn.angle);
+
                 const auto turn_instruction = turn.instruction;
 
                 if (turn_instruction.direction_modifier == guidance::DirectionModifier::UTurn)
                 {
                     distance += profile_properties.u_turn_penalty;
                 }
 
-                distance += turn_penalty;
+                // don't add turn penalty if it is not an actual turn. This heuristic is necessary
+                // since OSRM cannot handle looping roads/parallel roads
+                if (turn_instruction.type != guidance::TurnType::NoTurn)
+                    distance += turn_penalty;
```

Comment:
No. This actually only affects artificial turns due to failing to collapse.

if you have `a-b-a` as a road, we currently end up with two segments, even though it describes a circle. This introduces an artificial turn in between (at `b`).  The turn here will be undisturbed (no other roads) and on the same name. Only these turns are not penalised anymore.